### ISSUE TITLE
Simplify PouchDB ctor definitions

### ISSUE THREAD

**Issue by AGBrown** [AGBrown/pouchdb.d.ts#23]
Relates to #22 

Consider:

``` typescript
    /** The api module for the pouchdb callback pattern */
    module callback {
        /** The main pouchDB interface (callback pattern) */
        interface PouchDB extends api.db.Callback { }
    }
    /** The api module for the pouchdb promise pattern */
    module promise {
        /** The main pouchDB interface (promise pattern) */
        interface PouchDB extends api.db.Promisable { }
    }
    /** The api module for the pouchdb promise pattern (constructor only) */
    module thenable {
        /**
         * Special case class returned by the constructors of the promise api pouchDB.
         * Usually only a `pouchdb.promise.PouchDB` reference would be kept, assigned by
         * the `then` of the constructor.
         */
        interface PouchDB extends promise.PouchDB, async.Thenable<promise.PouchDB> { }
    }
```
1. `thenable.PouchDB` is a bad name (a real `thenable` does not have a `catch` anyway)
2. `thenable.PouchDB` is just there to be the return type for the promise-based constructors (i.e. to prevent `then` showing up on your pouchdb `db` variable)
3. `promise.PouchDB` is _exactly_  an `api.db.Promisable`
4. but `pouchdb.promise.PouchDB` is there because, ideally, you would be able to do `var db : pouchdb.PouchDB`. However we are splitting callable and promise interfaces to help end-users, so `var db : pouchdb.promise.PouchDB` is seemed marginally more explanatory (and has one less level of nesting) than `var db : pouchdb.api.db.Promisable`

@fredgalvao commented, rightly, that this structure is confusing. So ... how can we improve it?

Suggestions:
1. Rename `module promise` to `module promisable` in keeping with #22.
2. or rename `module thenable` to `module promisable` and remove `module promise`.
3. Or just move the interfaces out of `module db` and use those instead of `module callback` and `module promise`

I personally don't like this (which would be a vote against option 2):

``` typescript
    var dbp: pouchdb.api.db.Promisable;
     new PouchDB("name")
     .then(db => dbp = db);
```

compared to

``` typescript
    var dbp: pouchdb.promise.PouchDB;
     new PouchDB("name")
     .then(db => dbp = db);
```

Purely because `dbp` is a `PouchDB`, not a `Promisable`

Option 3 might be the best option. `module db` is a little superfluous - I only put it there to be clear where the main db interfaces were defined, but then ended up needing the `callback` and `promise` modules which are only there to improve the external naming.

**Comment by AGBrown** [comment]
Update from IRC chat with daleharvey:

TL;DR: pouchdb officially does not support the ctor returning a promise, so we should remove it. Updating this issue accordingly (cc @fredgalvao).

> <kbd>[14:10:54] &lt;AndyBrown></kbd> Did I read somewhere that creating a pouchdb is asynchronous, therefore `var db = new PouchDB()` is wrong, and the preferred method is to set db through the completion of a promise? Or did I just make that up?
> <kbd>[14:11:25] &lt;daleharvey> </kbd> its confusing, `var db = new PouchDB()` is the recommended way to setup a pouchdb instance
> <kbd>[14:12:12] &lt;AndyBrown> </kbd> thanks. So we _shouldn't_ set `db` using the `then` of a promise?
> <kbd>[14:13:19] &lt;daleharvey> </kbd> nope, thats the old way that we have moved away from
> <kbd>[14:12:26] &lt;daleharvey> </kbd> we do need an asyncronous setup, but we do it behind the scenes
> <kbd>[14:12:48] &lt;daleharvey> </kbd>  so **`new PouchDB().info()` will always work fine**
> <kbd>[14:13:42] &lt;AndyBrown> </kbd> moved away as in pre v3.4, or moved away as in v4.0?
> <kbd>[14:14:02] &lt;daleharvey> </kbd> yeh for a long time its not been recommended
> <kbd>[14:14:23] &lt;daleharvey> </kbd> still in there for backwards compat, but it has problems
> <kbd>[14:16:11] &lt;AndyBrown> </kbd> ... so with reference to #23: if I understood you properly, in fact we don't need to (or even, should absolutely not) have the object returned from the ctor extend promise? 
> <kbd>[14:16:53] &lt;daleharvey> </kbd> nope you **definitely shouldn't**, it causes all sorts of problems
> <kbd>[14:17:39] &lt;daleharvey> </kbd> I mean, the exact problems you are describing there are what we were having issues with
> <kbd>[14:17:58] &lt;daleharvey> </kbd> pouchdb is a promise to itself, just strange
> ...
> <kbd>[15:10:43] &lt;AndyBrown></kbd> daleharvey: if I remove promise from the object shape returned by the ctor, then I won't be able to do tests like: [test.basics.ts#L41-L45](https://github.com/AGBrown/pouchdb.d.ts/blob/master/tests/integration/test.basics.ts#L41-L45). Is it safe to just comment those out? If not I can probably find a workaround just for that test file (i.e. optionally add promise to the returned shape to support legacy use cases)
> <kbd>[15:11:13] &lt;AndyBrown></kbd> permalink to that code: [a6b2b045/test.basics.ts#L41-L45](https://github.com/AGBrown/pouchdb.d.ts/blob/a6b2b045/tests/integration/test.basics.ts#L41-L45)
> <kbd>[15:11:22] &lt;daleharvey></kbd> I would just take any tests / usage of the promise return out of there

**Comment by AGBrown** [comment]
Updated this issue, and the code (7ea6e816862ada4c79903f8cd0312efbe60ae5a0) following feedback from daleharvey. Comments from anyone? (cc @fredgalvao).

**Comment by fredgalvao** [comment]
That's awesome news! I didn't like that `thenable|promise` thing. I'm wondering why such things aren't removed from the code base on major version bumps though, as in 4.0.0 (which removed bluebird :cry: ).

Doesn't that imply that the callback parameter passed on to the constructor is also deprecated and should not be advised? (In the end it's just another way to deal with async behaviour on the db instance, so it should inherit the same reasoning...)
